<div class="h1-wrap">
    <h1 data-ui-scrollfix=79><span class="t_badge t_badge_icon"><span class="icon-info"></span></span>
        <translate>Information</translate>
    </h1>
</div>
<div data-scroll-spy="">
<aside class="t_sidenav" data-ui-scrollfix=79>
    <section>
        <ul class="t_subnav__items">
            <li data-spy="raw" class="t_subnav__items__item current">
                <a data-scrollto href="#raw">
                    <translate>Access Raw Data</translate>
                </a>
            </li>
            <li data-spy="methods" class="t_subnav__items__item">
                <a data-scrollto href="#methods">
                    <translate>Methods</translate>
                </a>
            </li>
            <li data-spy="tools" class="t_subnav__items__item">
                <a data-scrollto href="#tools">
                    <translate>Submitter Tools</translate>
                </a>
            </li>
        </ul>
    </section>
</aside>
<article>
    <section id="raw">
        <h3>
            <translate>Access Raw Data</translate>
            <a class="pilcrow" href="/info/raw"><i class="icon-pilcrow"></i></a>
        </h3>
        <h4><strong><translate>Public Repositories Hosting ICGC Data</translate></strong></h4>
        <p>
            <translate>
                ICGC member projects are required by ICGC policies to submit their raw sequencing reads,
                and other primary data, to controlled access public repositories. 
            </translate>
            <translate  has-markup>
                The official ICGC DCC-supported repository for <a href="http://www.ebi.ac.uk/ega/dacs/EGAC00001000010">ICGC sequencing reads</a>
                is the European Bioinformatics Institute's (EBI) <a href="http://www.ebi.ac.uk/ega">European Genome-phenome Archive (EGA)</a>.
            </translate>
        </p>

        <p>
            <translate has-markup>
                For the ICGC member projects which are part of the U.S <a href="http://cancergenome.nih.gov/">TCGA</a> project raw data will be hosted either at the <a href="http://tcga-data.nci.nih.gov/">TCGA DCC</a> or the <a href="https://cghub.ucsc.edu/ ">Cancer Genomics Hub (CGHub)</a>.
            </translate>
        </p>

        <h4><strong><translate>Applying for Access to Controlled-Access Datasets</translate></strong></h4>
        <p>
            <translate has-markup>
                Access to the ICGC sequencing datasets held at the EGA is granted by applying through 
                <a href="http://icgc.org/daco ">ICGC's Data Access and Compliance Office (DACO)</a>.
            </translate>
        </p>

        <h4><strong><translate>Overview of Available Raw Data</translate></strong></h4>
        <p>
            <translate has-markup>
                All controlled-access ICGC datasets held at EGA are associated with the <a href="http://icgc.org/daco">ICGC DACO</a>'s EGA DAC accession #:
            </translate> 
            <a href="https://www.ebi.ac.uk/ega/dacs/EGAC00001000010">EGAC00001000010</a>.
            
        </p>
    </section>
    <section id="methods">
        <h3><translate>Methods</translate><a class="pilcrow" href="/info/methods"><i class="icon-pilcrow"></i></a></h3>

        <h4><strong><translate>Gene model</translate></strong></h4>
        <p>
            <translate has-markup>
                The ICGC Data Portal uses gene models produced by Ensembl (<a href="http://www.ensembl.org">http://www.ensembl.org</a>).
            </translate>
        </p>
        <p>
            <translate has-markup>
                The canonical gene data (release 69) is available from Ensembl FTP site at 
                <a href="ftp://ftp.ensembl.org/pub/release-69/mysql/homo_sapiens_core_69_37">
                ftp://ftp.ensembl.org/pub/release-69/mysql/homo_sapiens_core_69_37</a>.
            </translate>
            <translate has-markup>
                We use a further-processed gene model file provided by <a href="http://snpeff.sourceforge.net/">snpEff</a> which is available for download from <a href="http://sourceforge.net/projects/snpeff/files/databases/v3_3/snpEff_v3_3_GRCh37.69.zip/download">
                http://sourceforge.net/projects/snpeff/files/databases/v3_3/snpEff_v3_3_GRCh37.69.zip/download</a>.
            </translate>
            <translate>
                This file is then transformed into a document-oriented data store for integration with other data types, such as, somatic mutations.
            </translate>
        </p>

        <h4><strong><translate>Gene consequence annotation</translate></strong></h4>
        <p>
            <translate>As part of the effort to enable cross cancer projects functional analysis, we perform standardized gene consequences annotation, using Sequence Ontology's controlled vocabulary regarding mutation-induced changes</translate> (<a href="www.sequenceontology.org">www.sequenceontology.org</a>).
            
            <translate has-markup>
                This type of annotation provides a list of predicted gene effects, such as a non-conservative amino acid substitution, but does not attempt to prioritize the changes by their functional impact.
            </translate> 
            <translate has-markup>
                Simple somatic mutation data submitted by member projects are reannotated using the variant annotation tool SnpEff (v3.3c, <a href="http://snpeff.sourceforge.net/">http://snpeff.sourceforge.net/</a>) recommended by the ICGC Mutation Consequences and Pathways (MUCOPA) Working Subgroup (Nature Methods, 2013, PMID:<a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=23900255 ">23900255</a>). 
            </translate>
        </p>

        <h4><strong><translate>Simple somatic mutation frequencies</translate></strong></h4>
        <p>
            <translate>
                In order to assess how often a particular simple somatic mutation (SSM) occurs in cancer patients, we establishes specific criteria to determine whether SSMs reported from different donors are considered to be the same.
            </translate>
            <translate>
                When all of these fields from two reported SSMs: chromosome, chromosome start, chromosome end, mutation type, DNA change, genome assembly version are identical, they are merged into one mutation entity.
            </translate>
            <translate>With this, we build a non-redundant collection of all simple somatic mutations.</translate>
            <translate>
                Each mutation entry in the collection is assigned a stable identifier for persisted referencing portal wide and across releases.
            </translate> 
            <translate>
                Information for linking these non-redundant SSMs to the reported ones is also kept, enabling mutation counts across donors.
            </translate>
        </p>
    </section>
    <section id="tools">
        <h3>
            <translate>Submitter Tools</translate>
            <a class="pilcrow" href="/info/tools"><i class="icon-pilcrow"></i></a>
        </h3>
        <p>
            <translate>
                This section provides instructions for ICGC projects on releasing data through the ICGC DCC and public repositories
            </translate>
        </p>

        <h4><strong><translate>Using the ICGC DCC Data Submission System (Beta)</translate></strong></h4>
        <p><translate>The ICGC DCC now supports data submissions through a web-based submission portal:</translate>
        <a href="https://submissions.dcc.icgc.org" target="_blank"><i class="icon-external-link"></i>https://submissions.dcc.icgc.org</a>.</p>

        <p>
            <translate has-markup>
                The submission system is currently accepting files using ICGC DCC data specification, which is available <a href="http://legacy-portal.dcc.icgc.org/pages/docs/dictionaries/latest" target="_blank"><i class="icon-external-link"></i>here</a> as an HTML document.
            </translate></p>

        <p>
            <translate has-markup>
                During the beta period, please refer to the 
                <a href="http://legacy-portal.dcc.icgc.org/pages/docs/icgc_dcc_data_submission_bulletin.pdf" target="_blank"><i class="icon-file"></i>ICGC Data Submission Bulletin</a>, and
                <a href="http://legacy-portal.dcc.icgc.org/pages/docs/dcc_submission_system_sop-beta.pdf" target="_blank"><i class="icon-file"></i>submission system software documentation</a> for instructions
                and important notes to assist you in using this system to release your project's data through the DCC data portal.
            </translate>
        </p>

        <p>
            <translate has-markup>
                Please contact <a href="mailto:dcc-support@icgc.org"><i class="icon-mail"></i>ICGC Data Coordination Centre</a> if you have any questions or comments about the data submission process.
            </translate>
        </p>

        <h4><strong><translate>Virtual machine for local data validation</translate></strong></h4>
        <p>
            <translate>
                The ICGC DCC offers a standalone virtual machine that allows you to locally validate ICGC submissions before uploading complete datasets to the online web-based data submission system.
            </translate>
            <translate>
                Please note that, as opposed to the DCC online submission system which is running on a computer cluster, this VM can be deployed only on a single server.
            </translate>
            <translate>
                It is not therefore suitable for validating entire datasets but rather meant for testing small subsets of data before submission to DCC.
            </translate>
            <translate has-markup>
                To use this virtual machine, please download the VirtualBox package from
                <a href="https://daw9opvs2y3v5.cloudfront.net/public/icgc-dcc-validator-1.10.ova" target="_blank">
                    <i class="icon-download"></i>here
                </a>.
            </translate>
            <translate has-markup>
                Instructions on how to use the VM can be found 
                <a href="http://legacy-portal.dcc.icgc.org/pages/docs/VirtualBoxforDCCv1.10.pdf" target="_blank">
                    <i class="icon-file"></i>here
                </a>.
            </translate>
        </p>
    </section>
</article>
</div>
