<h1>
    <img src="styles/images/PCAWG-icon.png" style="height:3.5rem; margin-left:2rem; vertical-align:bottom"></img>
    <span style="display:inline-block">
        <translate>PCAWG - Pancancer Analysis of Whole Genomes</translate>
    </span>
</h1>
<section class="no-border no-vert-padding">
    <div class="half pull-right" style="margin-bottom:1.5rem; margin-top:1rem; margin-left:2rem; margin-right: 2rem">

        <stacked height="250" data-title="{{'Donor Distribution by Primary Site' | translate}}"
            data-subtitle="{{ '[[projectCount]] projects and [[siteCount]] primary sites' | translate | subDelimiters:{projectCount: summary.projectCount, siteCount: summary.primarySiteCount} }}"
            data-y-label="{{'Donors' | translate}}" data-alternate-brightness="true" data-items="primarySites"
            data-is-loading="false" data-current-page="pcawg">
        </stacked>

        <section class="well" style="padding-top:0.5rem; padding-bottom:0.5rem">
            <div class="third text-center">
                <i class="icon-team" style="font-size:1.25rem"></i>
                <strong>{{summary.donorCount | number}}</strong>
                <translate>Donors</translate>
            </div>
            <div class="third text-center">
                <i class="icon-file" style="font-size:1.25rem"></i>
                <strong>{{summary.fileCount | number}}</strong>
                <translate>Files</translate>
            </div>
            <div class="third omega text-center">
                <i class="icon-database"
                    style="font-size:1.25rem"></i><strong>{{summary.totalFileSize | bytes}}</strong>
            </div>
        </section>

        <table class="table table-info">
            <thead>
                <th>
                    <translate>Data Type</translate>
                </th>
                <th>
                    <translate># Donors</translate>
                </th>
                <th>
                    <translate># Files</translate>
                </th>
                <th>
                    <translate>Format</translate>
                </th>
                <th>
                    <translate>Size</translate>
                </th>
            </thead>
            <tbody>
                <tr data-ng-repeat="type in pcawgDatatypes">
                    <td> {{type.uiName}}</td>
                    <td class="text-right">{{type.donorCount | number}}</td>
                    <td class="text-right">
                        <a href="{{ dataRepoUrl }}?filters={{type.filters}}">{{type.fileCount | number}}</a>
                    </td>
                    <td>{{ type.fileFormat.join(', ')}}</td>
                    <td class="text-right">{{type.fileSize | bytes}}</td>
                </tr>
            </tbody>
        </table>

        <span>
            <translate>Available data as of</translate> <a
                href="{{ dataRepoUrl }}?filters={{filters}}">{{indexDate | date}}</a>
        </span>
    </div>

    <p>
        <translate>
            The Pan-Cancer Analysis of Whole Genomes (PCAWG) study is an international collaboration to identify common
            patterns of mutation in more than 2,600 cancer whole genomes from the International Cancer Genome
            Consortium.
        </translate>
        <translate>
            Building upon previous work which examined cancer coding regions (Cancer Genome Atlas Research Network, The
            Cancer Genome Atlas Pan-Cancer analysis project, <span style="display:inline-block; text-indent:0"><a
                    href="https://www.nature.com/articles/ng.2764" target="_blank"><i
                        class="icon-external-link"></i>Nat. Genet. 2013 45:1113</a></span>, <span
                style="display:inline-block; text-indent:0"><a href="https://doi.org/10.1016/j.cell.2018.03.042"
                    target="_blank"><i class="icon-external-link"></i>Cell. 2018 Apr 5;173(2):283-285</a></span>), this
            project explored the nature and consequences of somatic and germline variations in both coding and
            non-coding regions, with specific emphasis on cis-regulatory sites, non-coding RNAs, and large-scale
            structural alterations.
        </translate>
    </p>
    <p>
        <translate>
            In order to facilitate the comparison among diverse tumor types, all tumor and matched normal genomes have
            been subjected to a uniform set of alignment and variant calling algorithms, and must pass a rigorous set of
            quality control tests.
        </translate>
        <translate>
            The research activities have been coordinated by a series of working groups comprising more than 700
            scientists.
        </translate>
    </p>
    <p>
        <translate>
            VCF-format files representing somatic variant calls for single-nucleotide variants, small indels, structural
            variants and copy number variants can be downloaded using the links in the table at the right.
        </translate>
        <translate>
            In addition, we provide aligned BAM files for download or cloud-based access via the <span
                style="display:inline-block; text-indent:0"><a href="https://cancercollaboratory.org/"
                    target="_blank"><i class="icon-external-link"></i>Cancer Genome Collaboratory</a></span> and <span
                style="display:inline-block; text-indent:0"><a href="https://dcc.icgc.org/icgc-in-the-cloud/aws"
                    target="_blank"><i class="icon-external-link"></i>Amazon Web Services</a></span>.
        </translate>
        <translate>
            Note that data sets that may contain germline SNPs are controlled tier and require credentials provided by
            the ICGC Data Access Committee and/or the TCGA dbGaP Data Access Committee.
        </translate>
        <translate>
            Open and controlled tier variants are also available for <span
                style="display:inline-block; text-indent:0"><a
                    href="https://dcc.icgc.org/search/m?filters=%7B%22mutation%22:%7B%22study%22:%7B%22is%22:%5B%22PCAWG%22%5D%7D%7D,%22donor%22:%7B%22studies%22:%7B%22is%22:%5B%22PCAWG%22%5D%7D%7D%7D&donors=%7B%22from%22:1%7D&mutations=%7B%22from%22:1%7D"
                    target="_blank"><i class="icon-external-link"></i>browsing and querying</a></span>.
        </translate>
        <translate>
            Detailed instructions for obtaining access to the controlled tier PCAWG data can be found in the <span
                style="display:inline-block; text-indent:0"><a href="https://docs.icgc.org/pcawg/data/"
                    target="_blank"><i class="icon-external-link"></i>DCC PCAWG documentation pages</a></span>.
        </translate>
    </p>
    <p>
        <translate>
            In addition, PCAWG working groups have generated substantial amounts of derived data, including donor
            clinical and histopathological data, subclonal reconstructions, purity and ploidy information, splice
            isoforms and mutational signatures.
        </translate>
        <translate>
            These data are described in individual PCAWG publications and can be downloaded from the <span
                style="display:inline-block; text-indent:0"><a href="https://dcc.icgc.org/releases/PCAWG"
                    target="_blank"><i class="icon-external-link"></i>ICGC PCAWG Data Release pages</a></span>.
        </translate>
    </p>
</section>
<section class="no-border no-vert-padding">
    <h3>
        <translate>Publication and Embargo Policy</translate>
    </h3>
    <p>
        <translate>
            The PanCancer Analysis of Whole Genomes (PCAWG) project primary and derived datasets are available for use
            by any researcher under the standard <span style="display:inline-block; text-indent:0"><a
                    href="https://icgc.org/index.php?q=icgc/goals-structure-policies-guidelines/e3-publication-policy"
                    target="_blank"><i class="icon-external-link"></i>ICGC publication policy terms</a></span> which
            allows for unrestricted use and publication provided that data is not used to identify donors, and that
            controlled tier data is kept confidential.
        </translate>
        <translate>
            The PCAWG publication embargo expired on July 25, 2019. When using this dataset, please cite reference #1
            below.
        </translate>
    </p>
</section>
<section class="no-border no-vert-padding">
    <p>
        <h3>
            <translate>Major Publications</translate>
        </h3>
        <ol>
            <li>The ICGC/TCGA Pan-Cancer Analysis of Whole Genomes Network. <strong>Pan-cancer analysis of whole
                    genomes</strong>. Nature (2020).</li>
            <li>Rheinbay, E. et al. <strong>Analyses of non-coding somatic drivers in 2,693 cancer whole
                    genomes</strong>. Nature (2020).</li>
            <li>PCAWG Transcriptome Core Group et al. <strong>Genomic basis of RNA alterations in cancer</strong>.
                Nature (2020).</li>
            <li>Li, Y. et al. <strong>Patterns of somatic structural variation in human cancer genomes</strong>. Nature
                (2020).</li>
            <li>Gerstung, M. et al. <strong>The evolutionary history of 2,658 cancers</strong>. Nature (2020).</li>
            <li>Alexandrov, L. B. et al. <strong>The Repertoire of Mutational Signatures in Human Cancer</strong>.
                Nature (2020).</li>
            <li>Phillips, M. et al. <strong>Of Clouds and Genomic Data Protection</strong>. Nature (2020).</li>
        </ol>
    </p>


</section>